splice_machine/src/main/java/com/splicemachine/derby/impl/storage/SpliceRegionAdmin.java

```diff
@@ -89,6 +92,290 @@
+    }
+
+    private static class  MoveRegionTask implements Callable<List<String>> {
+        private TxnView txn;
+        private String schemaName;
+        private String tableName;
+
+        public MoveRegionTask(TxnView txn,
+                              String schemaName,
+                              String tableName) {
+            this.txn = txn;
+            this.schemaName = schemaName;
+            this.tableName = tableName;
+        }
+
+        public List<String> call() {
+            List<String> result = Lists.newArrayList();
+            try (SpliceTransactionResourceImpl transactionResource = new SpliceTransactionResourceImpl()){
+                transactionResource.marshallTransaction(txn);
+                LanguageConnectionContext lcc = transactionResource.getLcc();
+                long conglomerateNumber = getConglomerateNumber(lcc, schemaName, tableName);
+                SpliceLogUtils.info(LOG, "schemaName=%s, tableName=%s, conglomerate=%d", schemaName, tableName, conglomerateNumber);
+                PartitionFactory partitionFactory = SIDriver.driver().getTableFactory();
+                Partition table = partitionFactory.getTable(Long.toString(conglomerateNumber));
+                List<Partition> partitions = table.subPartitions(new byte[0], new byte[0], true);
+                if (partitions.size() == 1) {
+                    PartitionServer ps = partitions.get(0).owningServer();
+                    String targetServer = getServerName(ps.getHostname(), ps.getPort(), ps.getStartupTimestamp());
+                    PartitionAdmin admin = SIDriver.driver().getTableFactory().getAdmin();
+                    TableDescriptor td = DataDictionaryUtils.getTableDescriptor(lcc, schemaName, tableName);
+                    ConglomerateDescriptorList cds = td.getConglomerateDescriptorList();
+                    for (ConglomerateDescriptor cd : cds) {
+                        long cn = cd.getConglomerateNumber();
+                        if (cn != conglomerateNumber) {
+                            table = partitionFactory.getTable(Long.toString(cn));
+                            partitions = table.subPartitions(new byte[0], new byte[0], true);
+                            for (Partition p : partitions) {
+                                String region = p.getEncodedName();
+                                String server = getServerName(p.owningServer().getHostname(),
+                                        p.owningServer().getPort(), p.owningServer().getStartupTimestamp());
+                                if (server.compareToIgnoreCase(targetServer) != 0) {
+                                    admin.move(region, targetServer);
+                                    result.add(String.format("Moved index %s from %s to %s", cd.getConglomerateName(),
+                                            server, targetServer));
+                                }
+                            }
+                        }
```

Comment:
I'm not sure if `admin.move` is blocking, or just scheduling a move.
if it is not blocking, ignore this comment.
however, if it is, potentially, this could move a lot of regions, and take a lot of time, and we would only see the first log message that it is trying to move some regions, but not how many, which might be a good indicator in case it is stuck/takes a long time.

i would consider this:
- collect a list of regions to be moved
- LOG the number of regions to be moved
- move regions, log done/todo

e.g.
```
               class SubTask {
                    String server, targetServer, region, conglomerateName;
                }

                List<SubTask> taskList = new ArrayList<>();
                for (ConglomerateDescriptor cd : cds) {
                    long cn = cd.getConglomerateNumber();
                    if (cn != conglomerateNumber) {
                        table = partitionFactory.getTable(Long.toString(cn));
                        partitions = table.subPartitions(new byte[0], new byte[0], true);
                        for (Partition p : partitions) {
                            String region = p.getEncodedName();
                            String server = getServerName(p.owningServer().getHostname(),
                                    p.owningServer().getPort(), p.owningServer().getStartupTimestamp());
                            if (server.compareToIgnoreCase(targetServer) != 0) {
                                SubTask st = new SubTask();
                                st.conglomerateName = cd.getConglomerateName();
                                st.region = region;
                                st.server = server;
                                st.targetServer = targetServer;
                                taskList.add(st);
                            }
                        }
                    }
                }
                SpliceLogUtils.info(LOG, "LocalizeIndicesTask schemaName=%s, tableName=%s, conglomerate=%d, will move "
                        + taskList.size() + " regions.", schemaName, tableName, conglomerateNumber);
                for(int i=0; i<taskList.size(); i++) {
                    SubTask task = taskList.get(i);
                    admin.move(task.region, task.targetServer);
                    String msg = String.format("Moved index %s from %s to %s", task.conglomerateName,
                            task.server, task.targetServer);
                    SpliceLogUtils.info(LOG, msg + ", " + i + "/" + taskList.size());
                    result.add(msg);
                }
```

Reply:
We only localize tables with a single region. Index is usually smaller than base table, so we only expect one region for each index. I wrote a loop here just to be conservative.